File: books/c-how-to-program/ch17/TimeClassCaseStudy.V3/Time.cpp
// Fig. 17.15: Time.cpp
// Time class member-function definitions.
#include <stdexcept>
#include "Time.h"
using namespace std;

Time::Time(int hr, int min, int sec)
{
    setTime(hr, min, sec);
}

void Time::setTime(int h, int m, int s)
{
    if (!((h >= 0 && h < 24) && (m >= 0 && m < 60) &&
        (s >= 0 && s < 60)))
        throw invalid_argument(
            "hour, minute and/or second was out of range");

    hour = h;
    minute = m;
    second = s;
}

int Time::getHour()
{
    return hour;
}

// POOR PRACTICE: Returning a reference to a private data member.
int &Time::badSetHour(int hh)
{
    if (!(hh >= 0 && hh < 24))
        throw invalid_argument("hour must be 0-23");

    hour = hh;

    return hour; // DANGEROUS reference return
}

File: books/java-how-to-program/src/examples/ch08/fig8_1_2_3/Time.java
// Fig. 8.1: Time.java
// Time class declaration maintains the time in 24-hour format.
package ch08.fig8_1_2_3;


public class Time {
    private int hour;
    private int minute;
    private int second;

    public void setTime(int hour, int minute, int second) {
        if (hour < 0 || hour >= 24 || minute < 0 || minute >= 60 ||
            second < 0 || second >= 60)
            throw new IllegalArgumentException(
                "hour, minute and/or second was out of range");
        
        this.hour = hour;
        this.minute = minute;
        this.second = second;
    }

    public String toUniversalString() {
        return String.format("%02d:%02d:%02d", hour, minute, second);
    }

    @Override
    public String toString() {
        return String.format("%d:%02d:%02d %s", 
        ((hour == 0 || hour == 12) ? 12 : hour % 12),
        minute, second, (hour < 12 ? "AM" : "PM"));
    }
    
}